### PR TITLE
Add support for `torch.compile` in `RGCNConv`

### LINKED ISSUE #8467: RGCNConv has multiple graph breaks
### 🚀 RGCNConv does barely have any speed up with torch_geometric.compile()

I have recently implemented torch_geometric.compile on my nn.module that mostly consists of the torch_geometric.nn.conv.rgcn_conv module. Analysing the graph breaks with torch._dynamo.explain() I have found that this module causes 6 graph breaks:

```
Graph Count: 7
Graph Break Count: 6
Op Count: 18
Break Reasons:
  Break Reason 1:
    Reason: call_function BuiltinVariable(int) [TensorVariable()] {}
    User Stack:
      <FrameSummary file /tmp/utke_s_pyg/tmpx31ev0wx.py, line 292 in forward>
  Break Reason 2:
    Reason: call_function BuiltinVariable(bool) [TensorVariable()] {}
    User Stack:
      <FrameSummary file /home/utke_s@WMGDS.WMG.WARWICK.AC.UK/miniconda3/envs/MADRL_py39/lib/python3.9/site-packages/torch_geometric/backend.py, line 53 in use_segment_matmul_heuristic>
Ops per Graph:
  Ops 1:
    <built-in method zeros of type object at 0x7fccf6ba7a40>
    <built-in method ones_like of type object at 0x7fccf6ba7a40>
  Ops 2:
    <built-in method tensor of type object at 0x7fccf6ba7a40>
    <built-in method tensor of type object at 0x7fccf6ba7a40>
    <built-in method tensor of type object at 0x7fccf6ba7a40>
    <built-in method tensor of type object at 0x7fccf6ba7a40>
    <built-in function sub>
    <built-in function truediv>
    <built-in function matmul>
    <built-in function ge>
  Ops 3:
    <built-in method zeros of type object at 0x7fccf6ba7a40>
    <built-in function getitem>
    <built-in function getitem>
    <built-in function truediv>
  Ops 4:
    <built-in method zeros of type object at 0x7fccf6ba7a40>
    <built-in function getitem>
    <built-in function getitem>
    <built-in function truediv>
  Ops 5:
  Ops 6:
  Ops 7:
```

Would it be possible to have a version of this module with reduced number of graph breaks? Unfortunately, I do not know enough about compilation and graph breaks to understand what causes the breaks, so any help on this would also be greatly appreciated!

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by akihironitta**
Fixes #8467.

### CODE REVIEW

File: test/nn/conv/test_rgcn_conv.py [hunk 1]
```diff
@@ -134,3 +139,35 @@ def test_rgcn_conv(cls, conf, device):
...
+    x2 = torch.randn(2, 16, device=device)
+    idx1 = torch.arange(4, device=device)
+    idx2 = torch.arange(2, device=device)
+    edge_index = torch.tensor([
+        [0, 1, 1, 2, 2, 3],
+        [0, 0, 1, 0, 1, 1],
+    ], device=device)
+    edge_type = torch.tensor([0, 1, 1, 0, 0, 1], device=device)
+    conv = cls(4, 32, 2, num_bases, num_blocks, aggr='sum').to(device)
+    explanation = torch._dynamo.explain(conv)(x1, edge_index, edge_type)
```
Comment (rusty1s) [inline test/nn/conv/test_rgcn_conv.py:161]: `explanation` doesn't seem to be used.

File: torch_geometric/nn/conv/rgcn_conv.py [hunk 2]
```diff
@@ -221,7 +221,9 @@ def forward(self, x: Union[OptTensor, Tuple[OptTensor, Tensor]],
 
         else:  # No regularization/Basis-decomposition ========================
 
-            use_segment_matmul = torch_geometric.backend.use_segment_matmul
+            use_segment_matmul: Optional[
+                bool] = torch_geometric.backend.use_segment_matmul
+            use_segment_matmul = not is_compiling()
```
Comment (rusty1s) [inline torch_geometric/nn/conv/rgcn_conv.py:226]: Looks like we override `use_segment_matmul` here.

File: torch_geometric/nn/conv/rgcn_conv.py [hunk 3]
```diff
@@ -372,3 +373,13 @@ def aggregate(self, inputs: Tensor, edge_type: Tensor, index: Tensor,
...
+
+
+# NOTE By updating the global flag directly in the RGCNConv.forward() without
+# allow_in_graph decorator, we see a graph break due to the following reason:
+# > Reason: store_attr
+# See https://pytorch.org/docs/main/torch.compiler_faq.html
+# #compiling-functions-besides-the-ones-which-are-supported-escape-hatch
+@allow_in_graph
+def disable_use_segment_matmul() -> None:
+    torch_geometric.backend.use_segment_matmul = False
```
Comment (rusty1s) [inline torch_geometric/nn/conv/rgcn_conv.py:385]: Wouldn't we need to reset it somewhere?

File: torch_geometric/nn/conv/rgcn_conv.py [hunk 4]
```diff
@@ -220,6 +221,8 @@ def forward(self, x: Union[OptTensor, Tuple[OptTensor, Tensor]],
                 out = out + h.contiguous().view(-1, self.out_channels)
 
         else:  # No regularization/Basis-decomposition ========================
+            if is_compiling():
+                disable_use_segment_matmul()
```
Comment (rusty1s) [inline torch_geometric/nn/conv/rgcn_conv.py:225]: Sorry for asking so many questions. What was the reason for the graph break in the first place? Computing the heuristic? Can't we just do
```
if is_compiling():
    use_segment_matmul = False
else:
    use_segment_matmul = torch_geometric.backend.use_segment_matmul
    if use_segment_matmul is None:
        pass
```
etc

**Comment by akihironitta** [comment]
I'll keep this PR on hold, as it makes more sense to first update the segment matmul heuristic (what @puririshi98 did in #8615). I see two graph breaks in the test case due to:

```
Break Reasons:
  Break Reason 1:
    Reason: call_function BuiltinVariable(int) [TensorVariable()] {}
    User Stack:
      <FrameSummary file /home/akihiro/work/github.com/pyg-team/pytorch_geometric/torch_geometric/nn/conv/rgcn_conv.py, line 235 in forward>
  Break Reason 2:
    Reason: call_function BuiltinVariable(bool) [TensorVariable()] {}
    User Stack:
      <FrameSummary file /home/akihiro/work/github.com/pyg-team/pytorch_geometric/torch_geometric/backend.py, line 55 in use_segment_matmul_heuristic>
```

**Pull request closed without merging**